tools/deploy_firmware.py

```diff
@@ -0,0 +1,287 @@
+        serial_path = "/dev/{}".format(self.name)
+        if not os.path.exists(serial_path):
+            raise ValueError("The serial device {} does not exist!\n".format(serial_path)
+                             + "Please make sure you have an appropriate udev rule for it.")
+        print("Copying image '{}' to the target".format(self.image))
+        try:
+            subprocess.check_output(["scp",
+                                     "{}/{}".format(self.artifacts_dir, self.image),
+                                     "{}:/tmp/{}".format(self.name, self.image)])
+        except subprocess.CalledProcessError as exc:
+            print("Failed to copy the image to the target:\n{}".format(exc.output))
+            raise exc
+        with serial.Serial(serial_path, self.baudrate) as ser:
+            print("Connecting to serial")
+            shell = pexpect.fdpexpect.fdspawn(ser, logfile=sys.stdout.buffer)
+            if not shell.isalive():
+                raise ValueError("Unable to connect to the serial device!")
+            print("Connected")
+            # The console might not be active yet:
+            shell.sendline("")
+            # make the shell prompt appear:
+            self.set_prompt(shell)
+            shell.expect(self.serial_prompt)
+            # Do the upgrade
+            shell.sendline("sysupgrade -v /tmp/{}".format(self.image))
+            # first give it 30 seconds, and fail early if the upgrade didn't start:
+            shell.expect(r"Performing system upgrade\.\.\.", timeout=30)
+            # now give it more time to apply the upgrade and reboot:
+            shell.expect(r"Rebooting system\.\.\.", timeout=300)
+            shell.expect("Please press Enter to activate this console", timeout=180)
+            # activate the console:
+            shell.sendline("")
+            self.set_prompt(shell)
+            shell.expect(self.serial_prompt)
+            shell.sendline("exit")
+        if not self.reach(attempts=10):
+            raise ValueError("The device was not reachable after the upgrade!")
+
+
+def main():
+    parser = argparse.ArgumentParser(prog=sys.argv[0],
+                                     description="""Update a prplWrt device, either through u-boot
+                                     or using sysupgrade, depending on the target device.""")
+    parser.add_argument('-d', '--device',
+                        help="""Device to upgrade. Currently supported targets are: netgear-rax40
+                        glinet-b1300 turris-omnia""", required=True)
+    parser.add_argument(
+        '-t',
+        '--target-name',
+        help="Name of the target to upgrade (make sure it's reachable through ssh).", required=True)
+
+    parser.add_argument(
+        '-i',
+        '--image',
+        help="Name of the image to use for the upgrade (should exist in the artifacts folder).",
+        required=True)
+
+    args = parser.parse_args()
+
+    if args.device == "netgear-rax40":
```

Comment:
perhaps compare to `args.device.lower()` to avoid capital letter issues in comparison.

Reply:
We use "netgear-rax40" everywhere as the name of the device/target, I don't see much advantage in making it case-insensitive. What use-case do you have in mind?